(dev/core#4490) Queues - Add setting and hook to pause/resume background-queues

eileenmcnaughton opened the pull request:
Overview
----------------------------------------

dev/core#4490 Add hook for `is_queue_active`


https://lab.civicrm.org/dev/core/-/issues/4490

Before
----------------------------------------
No way to 'tell the queue' to stop claiming items (eg.  because of server load) or because site is under maintenance

After
----------------------------------------
- Setting to put queues in maintenance mode (`queue_maintenance_mode`) which will prevent background processes from claiming items
- site can implement hook to prevent queue processing 

Technical Details
----------------------------------------
@totten @seamuslee001 how does this look compared to the discussion on the call ?

I updated the gitlab as per my understanding of the discussion https://lab.civicrm.org/dev/core/-/issues/4490

I did kinda think during discussion that the maintenance mode check would be done via a listener - so I just put the setting in

- I called the setting `queue_maintenance_mode` rather than just `maintenance_mode` as it felt like the latter might be over-promising?

- I haven't added to the UI as yet - I wasn't sure which page  - candidates are
  -  Settings - Debugging and Error Handling
  -  Misc (Undelete, PDFs, Limits, Logging, etc.)
  -  Settings - Cleanup Caches and Update Paths
  -  A brave new page - together with `enableBackgroundQueue` (& possibly some combo of  `prevNextBackend` `import_batch_size` `logging`, `environment`)

(If we went with the last option & it got tricksy it could be a follow up)


Comments
----------------------------------------

eileenmcnaughton committed:
Add hook for 'is_queue_active', 'queue_mainetenance_mode' setting

totten commented on CRM/Queue/Queue.php:
```diff
@@ -60,7 +62,17 @@ public function __construct($queueSpec) {
+      'queue_name' => $this->_name,
+      'queue_spec' => $this->queueSpec,
+    ]);
+    \Civi::dispatcher()->dispatch('is_queue_active', $event);
```
If you run `cv event /queue/`, you'll see some existing event-names look like:

```
[Event] civi.queue.check
[Event] civi.queue.runTask.start
[Event] hook_civicrm_queueRun_task
[Event] hook_civicrm_queueStatus
```

So maybe `civi.queue.isActive` or `hook_civicrm_queueActive`?

eileenmcnaughton replied:
@totten do you have a sense of which of the above it is closest to logically? My instinct is perhapes queueStatus

eileenmcnaughton replied:
@totten I've updated to the first suggestion in the spirit of 'pick one any one'

totten replied:
Yeah, it just depends how public one wants to make it. The first one (`civi.queue.isActive`; less public) sounds good.

Weird thing - Github UI isn't showing the update. (Well, it shows a force-push, but the main diff hasn't changed.) Maybe if we "Resolve conservation" it will refresh?

totten replied:
Nope that, didn't help. Weird.

eileenmcnaughton replied:
So I see 

![image](https://github.com/civicrm/civicrm-core/assets/336308/8182da12-edc3-40e3-a5e1-c24c667f1240)

eileenmcnaughton commented on CRM/Queue/Queue.php:
```diff
@@ -60,7 +62,17 @@ public function __construct($queueSpec) {
   public function isActive(): bool {
     $status = CRM_Core_DAO::getFieldValue('CRM_Queue_DAO_Queue', $this->_name, 'status', 'name', TRUE);
-    // Note: In the future, we may want to incorporate other data (like maintenance-mode or upgrade-status) in deciding active queues.
+    if ($status === 'active' && \Civi::settings()->get('civi.queue.isActive')) {
```
Oh weird @totten I see 

![image](https://github.com/civicrm/civicrm-core/assets/336308/799475f2-cac1-4451-900e-99e35b1b8089)

totten replied:
Yep, it looked better for me now too.

mattwire commented on settings/Core.setting.php:
```diff
@@ -1201,4 +1201,16 @@
     'description' => ts('How long should HTTP requests through Guzzle application run for in seconds'),
     'help_text' => ts('Set the number of seconds http requests should run for before terminating'),
   ],
+  'queue_maintenance_mode' => [
+    'name' => 'queue_maintenance_mode',
+    'type' => 'Boolean',
+    'default' => FALSE,
+    'html_type' => 'yesno',
+    'add' => '5.67',
+    'title' => ts('Is Queue processing in maintenance mode?'),
+    'is_domain' => 1,
+    'is_contact' => 0,
+    'description' => ts('If enabled, CiviCRM will not process background queues.'),
+    'help_text' => ts('This setting will only affect sites that have background queue processing enabled (eg. coworker)'),
+  ],
```
@eileenmcnaughton This doesn't seem to actually be used anywhere? Also, does it really need to be called "maintenance mode" when it's actually just a setting to disable queue processing that might be used in various circumstances? I'm guessing it needs to be a setting because it's being read by the external queue processor?
What about something like "queue_disable" and then title "Disable background queue processing".

eileenmcnaughton replied:
@mattwire argh - it's not being called cos I screwed it up on fixing the listener name ....  I'll fix that. I'm OK with queue_disable

eileenmcnaughton replied:
@mattwire ok - updated

mattwire replied:
Great! I'm happy with that now

eileenmcnaughton replied:
cool thanks @mattwire - that also explains why @totten couldn't see my change earlier - I put the listener name in the setting.get call. 

Also - note sure if you saw it but we had a discussion earlier today & notes/recording/transcript are in product maintenance on chat

totten committed:
Use subsystem prefix on new setting

eileenmcnaughton commented:
Just confirming I'm +1 on your change @totten so merge when ready

totten committed:
Queue::isActive() - Use the latest value of the setting

totten committed:
The Battle of the Synonyms was put on-hold.

With the recent update, it fires a hook, and the hook-data describes the
queue-status as "deferred".

But the setting which drives it is called "queue_suspended".

Which word is better?  On reflection...  neither!

In CiviMail, there's some existing functionality where you can "pause" a
mail-blast.  And it's basically the same thing: we keep the queue-data, but
we tell the workers to temporarily refrain from execution.  And that's fine.
"Pausing" sounds less scary...

totten commented on CRM/Queue/Queue.php:
```diff
@@ -60,7 +62,17 @@ public function __construct($queueSpec) {
   public function isActive(): bool {
     $status = CRM_Core_DAO::getFieldValue('CRM_Queue_DAO_Queue', $this->_name, 'status', 'name', TRUE);
-    // Note: In the future, we may want to incorporate other data (like maintenance-mode or upgrade-status) in deciding active queues.
+    if ($status === 'active' && \Civi::settings()->get('queue_suspended')) {
```
@eileenmcnaughton I'm not certain if this is an issue for you or not, but I did some `r-run`, and it didn't quite work the way I expected

The issue is that `Civi::settings()` is designed to cache the setting. But we have multiple concurrent workers (pipe-sessions); each worker has a separate cache. Eventually [workers expire](https://lab.civicrm.org/dev/coworker/-/blob/v0.2/src/Configuration.php?ref_type=tags#L22-55) -- so eventually, the cache is reset. But I would expect the flag to be more immediate.

Pushing up a patch...

eileenmcnaughton replied:
thanks @totten

totten commented:
OK, this works for me now in `r-run`. We'll need new CI run, and the latest push could get some feedback. But overall, I think it's ready. Flagged `merge ready`.

eileenmcnaughton commented:
Cool those changes look good
